### PR TITLE
[Core] Log output from different jobs to different drivers. 

### PR DESCRIPTION
**Pull request by wuisawesome**
## Why are these changes needed?

This is a first step to handling ensuring that only output from the relevant job is sent to each driver. This PR will make each worker write its own output to a file titled `worker-{worker-id}.out` (which it already does), but use produced output from jobs will now be written to a file titled `worker-{worker-id}-{job-id}.out`. stderr will also be redirected accordingly.

## Related issue number

First of 2 steps towards closing #4029.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested (please justify below)

No tests here. In some ways this is just a non-trivial refactoring in the sense that it shouldn't effect end behavior, thus it's sufficient to not break anything. When behavior changes (i.e. next PR where output is actually sent to specific drivers only), test cases will be included.

### CODE REVIEW

File: python/ray/_raylet.pyx [hunk 1]
```diff
@@ -700,6 +772,12 @@ cdef class CoreWorker:
         return ActorID(
             CCoreWorkerProcess.GetCoreWorker().GetActorId().Binary())
 
+    def get_stdout_file(self):
+        return CCoreWorkerProcess.GetCoreWorker().GetStdoutFile()
```
[the remark this answers is not in the document]
Comment (wuisawesome) [reply python/ray/_raylet.pyx:807]: This is the worker's stdout (not associated with any job). I think what you're describing is `get_job_redirected_log_file`

**Comment by rkooo567** [comment]
@wuisawesome Can you also post microbenchmark result before/after this PR? You can do `ray microbenchmark`

File: src/ray/util/util.cc [hunk 2]
```diff
@@ -323,3 +322,21 @@ std::string CreateCommandLine(const std::vector<std::string> &args,
...
+}
+
+void flush_err() {
+  // Flush c++ buffer
+  std::cerr << std::flush;
+  // Flush c streaming buffer
+  // TODO(alex): If we don't actually use the C streaming api then we may not
+  // need this line at all
+  fflush(stderr);
+}
```
Comment (ericl) [inline src/ray/util/util.cc:342]: Let's revert the C++ changes.

File: python/ray/utils.py [hunk 3]
```diff
@@ -390,6 +390,36 @@ def setup_logger(logging_level, logging_format):
...
+    # named-pipes.
+    f = open_log(path)
+    # Check to see if we're creating this file. No one else should ever write
+    # to this file, so we don't have to worry about TOCTOU.
+    if f.tell() == 0:
+        # This should always be the first message to appear in the worker's
+        # stdout and stderr log files. The string "Ray worker pid:" is
+        # parsed in the log monitor process.
+        print("Ray worker pid: {}".format(worker_pid), file=f)
+        f.flush()
```
Comment (ericl) [inline python/ray/utils.py:419]: Let's avoid flushing in the first version (let's see).

**Comment by wuisawesome** [comment]
Here's a run of `ray microbenchmark`:

```single client get calls per second 9229.64 +- 22.38
single client put calls per second 5456.76 +- 17.89
single client put gigabytes per second 5.25 +- 0.43
multi client put calls per second 11584.16 +- 109.5
multi client put gigabytes per second 14.66 +- 6.69
single client tasks sync per second 1308.76 +- 47.58
single client tasks async per second 16260.1 +- 208.94
multi client tasks async per second 49602.8 +- 2675.45
1:1 actor calls sync per second 2008.02 +- 32.31
1:1 actor calls async per second 7037.04 +- 191.16
1:1 actor calls concurrent per second 6264.77 +- 232.32
1:n actor calls async per second 10879.23 +- 155.75
n:n actor calls async per second 36125.86 +- 548.97
n:n actor calls with arg async per second 13114.92 +- 142.85
1:1 async-actor calls sync per second 1277.12 +- 31.9
1:1 async-actor calls async per second 3755.65 +- 47.24
1:1 async-actor calls with args async per second 2336.81 +- 36.62
1:n async-actor calls async per second 9452.64 +- 86.85
n:n async-actor calls async per second 28349.32 +- 243.23
```

compared to master:

```
single client get calls per second 9550.3 +- 349.69
single client put calls per second 5412.16 +- 94.52
single client put gigabytes per second 5.31 +- 0.42
multi client put calls per second 11643.22 +- 194.13
multi client put gigabytes per second 13.58 +- 5.85
single client tasks sync per second 1447.68 +- 36.93
single client tasks async per second 17018.14 +- 604.89
multi client tasks async per second 47783.35 +- 2194.71
1:1 actor calls sync per second 2346.24 +- 126.24
1:1 actor calls async per second 6902.03 +- 180.9
1:1 actor calls concurrent per second 6227.98 +- 148.66
1:n actor calls async per second 10762.19 +- 219.83
n:n actor calls async per second 36035.61 +- 469.98
n:n actor calls with arg async per second 13092.93 +- 318.03
1:1 async-actor calls sync per second 1221.29 +- 34.01
1:1 async-actor calls async per second 3779.37 +- 94.78
1:1 async-actor calls with args async per second 2576.19 +- 8.77
1:n async-actor calls async per second 9222.4 +- 207.06
n:n async-actor calls async per second 28158.28 +- 584.81
```

File: python/ray/_raylet.pyx [hunk 4]
```diff
@@ -184,6 +184,96 @@ def compute_task_id(ObjectID object_id):
...
+
+        # We must close the original file descriptor to avoid leaking file
+        # descriptors.
+        stdout_file.close()
+
+        # We also manually set sys.stdout and sys.stderr because that seems to
+        # have an affect on the output buffering. Without doing this, stdout
+        # and stderr are heavily buffered resulting in seemingly lost logging
+        # statements. We never want to the stdout file descriptor, dup2 will
+        # close it when necessary and we don't want python's GC to close it.
```
Comment (mehrdadn) [inline python/ray/_raylet.pyx:235]: There's a lot going on here and I think some of it is confusing. Setting a field like `sys.stdout` doesn't inherently affect buffering (buffering is a property of the stream itself). Furthermore, the default stdio buffering behavior is affected by whether Python is started via `python -u`, so that's one way to disable buffering for spawned processes (such as workers). If you're trying to do this for the driver, though... I feel like it's the caller's responsibility to manage stdio, and to ensure they're redirected before the process is spawned. It seems a bit unexpected to pull the rug out from underneath the caller, since the caller (which may be the user) might deliberately set up stdio in some particular way. It might be good to discuss the approach to see if there might be a better way?

File: python/ray/utils.py [hunk 5]
```diff
@@ -390,6 +390,35 @@ def setup_logger(logging_level, logging_format):
     logger.propagate = False
 
 
+def open_log(path, **kwargs):
+    kwargs.setdefault("buffering", 1)
+    kwargs.setdefault("mode", "a")
+    return open(path, **kwargs)
```
Comment (mehrdadn) [inline python/ray/utils.py:396]: I'd suggest `open(..., encoding="utf-8")` because on Windows this doesn't default to UTF-8 (which I presume the log files are assumed to be).

File: python/ray/_raylet.pyx [hunk 6]
```diff
@@ -184,6 +184,96 @@ def compute_task_id(ObjectID object_id):
     return TaskID(object_id.native().TaskId().Binary())
 
 
+def setup_logging(stdout_name, stderr_name):
+    """Sets up logging for the current worker, creating the file and flushing
+    buffers as is necessary.
```
[the remark this answers is not in the document]
Comment (wuisawesome) [reply python/ray/_raylet.pyx:189]: was originally there in order to call out to flush c++ buffers, but yes, I'll move it

File: python/ray/_raylet.pyx [hunk 7]
```diff
@@ -184,6 +184,96 @@ def compute_task_id(ObjectID object_id):
...
+        (tuple) The absolute paths of the files that stdout and stderr will be
+    written to.
+    """
+    stdout_path = ""
+    stderr_path = ""
+    worker_pid = os.getpid()
+
+    cdef extern from "ray/util/util.h":
+        void flush_out()
+        void flush_err()
```
[the remark this answers is not in the document]
Comment (wuisawesome) [reply python/ray/_raylet.pyx:205]: Not anymore :) good catch

File: python/ray/_raylet.pyx [hunk 8]
```diff
@@ -184,6 +184,96 @@ def compute_task_id(ObjectID object_id):
...
+
+    if stdout_name:
```
[the remark this answers is not in the document]
Comment (wuisawesome) [reply python/ray/_raylet.pyx:207]: maybe it would look better if I didn't put equivalent comments in the stderr part...

File: python/ray/_raylet.pyx [hunk 9]
```diff
@@ -184,6 +184,96 @@ def compute_task_id(ObjectID object_id):
...
+        # Line-buffer the output (mode 1).
+        stdout_file = open_worker_log(stdout_name, worker_pid)
+
+        # Before redirecting stdout we need to make sure userspace buffers
+        # are all flushed.
+        # Flush python's userspace buffers
```
[the remark this answers is not in the document]
Comment (wuisawesome) [reply python/ray/_raylet.pyx:213]: I think this is a useful comment because it's important to know why we're doing an expensive operation. For example, if in the future we always start workers with `python -u` then we shouldn't have to guess if removing this line will have any other effect.

[the remark this answers is not in the document]
Comment (mehrdadn) [reply python/ray/_raylet.pyx:213]: I'm actually unclear on this⁠—what actually does happen in that case? Normally `python -u`'s job is to make `sys.stdout.flush()` unnecessary (and similar for `stderr` etc.). If the `flush()` would still be necessary in that case, it might be good to explain why in the comment? Or if it isn't, then it might be useful to leave a note that we can remove it if we switch to `python -u` (which I think we should btw)?

[the remark this answers is not in the document]
Comment (wuisawesome) [reply python/ray/_raylet.pyx:213]: I don't believe we'd have to flush if stdout was guaranteed to be unbuffered. There's also a variety of ways of making sure that stdout isn't buffered, so I'd prefer to state the rationale since we can't provide comprehensive instructions. 

Switching to an unbuffered output might make sense, but it I think it should be in it's own PR since it would have its own independent performance characteristics.

[the remark this answers is not in the document]
Comment (mehrdadn) [reply python/ray/_raylet.pyx:213]: Sounds good!

File: python/ray/utils.py (hunk 5 above)
Comment (wuisawesome) [reply python/ray/utils.py:396]: Will do, has this not come up before though?

Comment (mehrdadn) [reply python/ray/utils.py:396]: It will once people start running Ray on Windows and non-ASCII characters get emitted into the log :-)

File: python/ray/_raylet.pyx [hunk 10]
```diff
@@ -184,6 +184,96 @@ def compute_task_id(ObjectID object_id):
...
+        # We also manually set sys.stdout and sys.stderr because that seems to
+        # have an affect on the output buffering. Without doing this, stdout
+        # and stderr are heavily buffered resulting in seemingly lost logging
+        # statements. We never want to the stdout file descriptor, dup2 will
+        # close it when necessary and we don't want python's GC to close it.
+        sys.stdout = open_log(stdout_fileno, closefd=False)
+
+        stdout_path = os.path.abspath(stdout_file.name)
+
+    # The stderr case should be analogous to the stdout case
```
[the remark this answers is not in the document]
Comment (wuisawesome) [reply python/ray/_raylet.pyx:240]: setting `sys.stderr` vs `sys.stdout` is a little awkward and saves like 3 lines of code.

[the remark this answers is not in the document]
Comment (mehrdadn) [reply python/ray/_raylet.pyx:240]: The issue with duplication isn't just with the LOC (although with the comments, these are entire screenfuls even right now), but other things too, e.g. like the fact that the synchronization between them is brittle. They get out of sync very quickly, and it becomes difficult to check if they're still semantically equivalent. In general you only want to duplicate code where the similarity is merely a coincidence, and the evolution likely to be unrelated—this happens on occasion (especially for small blocks of code), but it's rather rare, and not the case here. In a situation like this, where the two streams should be treated more or less identically, having the same code path for both helps make the code easier to maintain.

File: python/ray/_raylet.pyx (hunk 7 above)
[the remark this answers is not in the document]
Comment (mehrdadn) [reply python/ray/_raylet.pyx:205]: Cool! (Feel free to resolve these yourself after pushing btw.)

File: python/ray/_raylet.pyx [hunk 11]
```diff
@@ -184,6 +184,96 @@ def compute_task_id(ObjectID object_id):
...
+        # TODO (Alex): Flush the c/c++ userspasce buffers if necessary.
+        # `fflush(stdout); cout.flush();`
+
+        stdout_fileno = sys.stdout.fileno()
+
+        # Redirect stdout at the file descriptor level. If we simply set
+        # sys.stdout, then logging from C++ can fail to be
+        # redirected. Note that dup2 will automatically close the old file
+        # descriptor before overriding it.
+        os.dup2(stdout_file.fileno(), stdout_fileno)
```
[the remark this answers is not in the document]
Comment (wuisawesome) [reply python/ray/_raylet.pyx:225]: This one can't error because we're the ones opening the file.

[the remark this answers is not in the document]
Comment (mehrdadn) [reply python/ray/_raylet.pyx:225]: Oh I see. Maybe put this as a postcondition of `open_worker_log` (doc + `assert`)? Right now it just says the return value is a "file-like object" which seems to imply it doesn't necessarily need to be bound to a file descriptor.

[the remark this answers is not in the document]
Comment (wuisawesome) [reply python/ray/_raylet.pyx:225]: I say "file-like object" because I think in the long term we should actually replace this with named pipes.

[the remark this answers is not in the document]
Comment (mehrdadn) [reply python/ray/_raylet.pyx:225]: Right, that make sense. Just something to indicate it should have a `fileno()` would be nice since right now you can't tell just by looking at the function; it's an implicit requirement.

File: python/ray/_raylet.pyx [hunk 12]
```diff
@@ -184,6 +184,96 @@ def compute_task_id(ObjectID object_id):
...
+
+        # Before redirecting stdout we need to make sure userspace buffers
+        # are all flushed.
+        # Flush python's userspace buffers
+        sys.stdout.flush()
+
+        # TODO (Alex): Flush the c/c++ userspasce buffers if necessary.
+        # `fflush(stdout); cout.flush();`
+
+        stdout_fileno = sys.stdout.fileno()
```
[the remark this answers is not in the document]
Comment (wuisawesome) [reply python/ray/_raylet.pyx:219]: Here's the thing:

(1) This is the existing code behavior (and there doesn't seem to be a problem).
(2) We control how workers are started. It would be quite hard to start a worker without calling `ray start`.
(3) The underlying issue is that IDLE has a poorly design stdout object (i.e. it doesn't follow the Liskov Substitution Principle). 

I don't think need to (or even should) introduce additional code to support bad practices, which no one uses, in order to make the failure messages a little nicer (either way we're going to fail on startup anyways).

File: python/ray/_raylet.pyx (hunk 4 above)
Comment (wuisawesome) [reply python/ray/_raylet.pyx:235]: I haven't looked super closely into why this is necessary. This is refactored in from [here](https://github.com/ray-project/ray/blob/6c49c01837ef9d83b998a2ec1a269a1a6767fc66/python/ray/worker.py#L1172) (perhaps @robertnishihara remembers). 

Existing behavior is we want stdout to have `open(..., buffering=1)` so we probably want to do that at least once.

File: python/ray/_raylet.pyx (hunk 12 above)
[the remark this answers is not in the document]
Comment (mehrdadn) [reply python/ray/_raylet.pyx:219]: I actually don't think it's an LSP violation; it's consistent with the file protocol, and I'm not aware of any extra requirements on `stdout` et al. :-) [`IOBase`](https://docs.python.org/3/library/io.html#io.IOBase.fileno) does say this:

> Return the underlying file descriptor (an integer) of the stream if it exists. An `OSError` is raised if the IO object does not use a file descriptor.

I'm not really worried about whether it's bad practice or not (I can argue both ways), but I'm worried about the UX. Programs often have to work around environment idiosyncrasies to provide a better experience.

As to why this hasn't come up yet, I'm unclear on whether this also applies to the driver or not. If it does, and it hasn't come up, my guess is it's most likely due to the fact that IDLE tends to be used on Windows, which we didn't previously support. I think Linux/Mac users rarely run IDLE and instead type `python` directly in their terminals.

That said, if you feel strongly about it regardless, I'll probably just end up changing it myself later, so it's not a big deal either way. Just wanted to point it out because having various packages break in IDLE due to this issue is something I encounter fairly regularly on Windows.

File: python/ray/_raylet.pyx (hunk 4 above)
Comment (mehrdadn) [reply python/ray/_raylet.pyx:235]: Ohh I see. We can probably ignore changing the approach for this PR then, it didn't hit me it was what the code was already doing in `worker.py`.

File: python/ray/_raylet.pyx (hunk 12 above)
[the remark this answers is not in the document]
Comment (wuisawesome) [reply python/ray/_raylet.pyx:219]: This should be worker only. I don't have a windows machine (or even IDLE) readily available for testing, but if you can currently run `ray.init()` in IDLE then this won't change that functionality. If you REALLY want I can add a try/catch, but I believe the correct behavior should be to fail and propagate an error message up if someone tries to start a worker in IDLE.

[the remark this answers is not in the document]
Comment (mehrdadn) [reply python/ray/_raylet.pyx:219]: Ah ok, I guess feel free to keep it as is then. Maybe it won't come up. I can fix it if it comes up haha.

File: python/ray/node.py [hunk 13]
```diff
@@ -639,10 +652,47 @@ def start_raylet(self, use_valgrind=False, use_profiler=False):
         assert ray_constants.PROCESS_TYPE_RAYLET not in self.all_processes
         self.all_processes[ray_constants.PROCESS_TYPE_RAYLET] = [process_info]
 
+    def get_job_redirected_log_file(self, worker_id, job_id):
```
[the remark this answers is not in the document]
Comment (wuisawesome) [reply python/ray/node.py:655]: Agreed that there's a lot of files, but we generate a lot of logs. It's complicated because worker, job, redis, raylet, etc logs all share parts of this code path.

[the remark this answers is not in the document]
Comment (mehrdadn) [reply python/ray/node.py:655]: Sorry, I didn't mean a lot of files, I meant a lot of **functions** (in the code) to open files, and it's unclear which one is responsible for what (or why opening a log requires so many functions). Ideally there would be 1 function responsible for opening any log file, though that's probably unrealistic. But can we simplify it from the 6+ we have now?

File: python/ray/_raylet.pyx [hunk 14]
```diff
@@ -184,6 +184,96 @@ def compute_task_id(ObjectID object_id):
...
+
+        # Redirect stdout at the file descriptor level. If we simply set
+        # sys.stdout, then logging from C++ can fail to be
+        # redirected. Note that dup2 will automatically close the old file
+        # descriptor before overriding it.
+        os.dup2(stdout_file.fileno(), stdout_fileno)
+
+        # We must close the original file descriptor to avoid leaking file
+        # descriptors.
```
[the remark this answers is not in the document]
Comment (wuisawesome) [reply python/ray/_raylet.pyx]: Ah you're right, GC should handle that.

File: python/ray/_raylet.pyx [hunk 15]
```diff
@@ -184,6 +184,96 @@ def compute_task_id(ObjectID object_id):
...
+        stdout_file.close()
+
+        # We also manually set sys.stdout and sys.stderr because that seems to
+        # have an affect on the output buffering. Without doing this, stdout
+        # and stderr are heavily buffered resulting in seemingly lost logging
+        # statements. We never want to the stdout file descriptor, dup2 will
+        # close it when necessary and we don't want python's GC to close it.
+        sys.stdout = open_log(stdout_fileno, closefd=False)
+
+        stdout_path = os.path.abspath(stdout_file.name)
```
[the remark this answers is not in the document]
Comment (wuisawesome) [reply python/ray/_raylet.pyx]: I think the missing here is that the dependency isn't on the current working directory, it's on the temp file service. Relative names _should_ be provided, but the absolute path isn't just `$(pwd)/$FILE`. Returning the absolute path means that the other helper processes (like log monitor) don't need the extra dependency.

File: python/ray/_raylet.pyx [hunk 16]
```diff
@@ -306,6 +306,19 @@ cdef prepare_args(
                         core_worker.put_serialized_object(serialized_arg)))))
 
 
+def switch_worker_log(worker, next_job_id):
```
Comment (ericl) [inline python/ray/_raylet.pyx:309]:
```suggestion
def switch_worker_log_if_needed(worker, next_job_id):
```

File: python/ray/_raylet.pyx [hunk 17]
```diff
@@ -442,6 +455,7 @@ cdef execute_task(
             with core_worker.profile_event(b"task:execute"):
                 task_exception = True
                 try:
+                    switch_worker_log(worker, job_id)
```
Comment (ericl) [inline python/ray/_raylet.pyx:458]:
```suggestion
                    switch_worker_log_if_needed(worker, job_id)
```

File: python/ray/node.py [hunk 18]
```diff
@@ -383,14 +383,14 @@ def _make_inc_temp(self, suffix="", prefix="", directory_name=None):
         raise FileExistsError(errno.EEXIST,
                               "No usable temporary filename found")
 
-    def new_log_files(self, name):
+    def get_log_file_names(self, name, unique=False):
         """Generate partially randomized filenames for log files.
 
         Args:
             name (str): descriptive string for this log file.
```
Comment (ericl) [inline python/ray/node.py:390]: Document the 'unique' arg.

File: python/ray/node.py [hunk 19]
```diff
@@ -401,14 +401,27 @@ def new_log_files(self, name):
...
+        """Creates partially randomized filenames for log files.
+
+        Args:
+            name (str): descriptive string for this log file.
+
+        Returns:
+            A tuple of two files for redirecting (stdout, stderr).
+        """
+        log_stdout, log_stderr = self.get_log_file_names(name, unique=True)
+        return open_log(log_stdout), open_log(log_stderr)
```
Comment (ericl) [inline python/ray/node.py:424]: Consider removing the function and inlining the logic into callers.

File: python/ray/node.py (hunk 13 above)
Comment (ericl) [inline python/ray/node.py:655]: Can we start annotating the types here?

File: python/ray/node.py [hunk 20]
```diff
@@ -639,10 +652,47 @@ def start_raylet(self, use_valgrind=False, use_profiler=False):
...
+            worker_id (bytes): A byte representation of the worker id.
+            job_id (bytes): A byte representation of the job id.
+
+        Returns:
+            (tuple) The stdout and stderr file names that the job should be
+        redirected to.
+        """
+        redirect_output = self._ray_params.redirect_output
+
+        if redirect_output is None:
```
Comment (ericl) [inline python/ray/node.py:669]: Can we remove this?

File: python/ray/node.py [hunk 21]
```diff
@@ -639,10 +652,47 @@ def start_raylet(self, use_valgrind=False, use_profiler=False):
...
+                ray.utils.binary_to_hex(worker_id),
+                ray.utils.binary_to_hex(job_id)),
+            unique=False))
+        return worker_stdout_file, worker_stderr_file
+
     def new_worker_redirected_log_file(self, worker_id):
-        """Create new logging files for workers to redirect its output."""
-        worker_stdout_file, worker_stderr_file = (
-            self.new_log_files("worker-" + ray.utils.binary_to_hex(worker_id)))
+        """Determines (but does not create) logging files for workers to
```
Comment (ericl) [inline python/ray/node.py:684]: Can we combine this with the above function? Have an optional current_job_id=None.

File: python/ray/utils.py [hunk 22]
```diff
@@ -390,6 +390,36 @@ def setup_logger(logging_level, logging_format):
     logger.propagate = False
 
 
+def open_log(path, **kwargs):
+    kwargs.setdefault("buffering", 1)
+    kwargs.setdefault("mode", "a")
+    kwargs.setdefault("encoding", "utf-8")
```
Comment (ericl) [inline python/ray/utils.py:396]: We don't anywhere else right? If this is actually required can it be in another PR?

Comment (ericl) [reply python/ray/utils.py:396]: I just don't want to one-off change this particular log file encoding and not the others.

Comment (ericl) [reply python/ray/utils.py:396]: Actually can we just remove this file and call `open(file, "a", buffering=1)` directly? This is same as the other comment to inline thin wrappers.

File: python/ray/utils.py [hunk 23]
```diff
@@ -390,6 +390,36 @@ def setup_logger(logging_level, logging_format):
...
+    return open(path, **kwargs)
+
+
+def open_worker_log(path, worker_pid):
```
Comment (ericl) [inline python/ray/utils.py:400]:
```suggestion
def create_and_init_new_worker_log(path, worker_pid):
```

File: python/ray/worker.py [hunk 24]
```diff
@@ -891,6 +887,71 @@ def custom_excepthook(type, value, tb):
...
 
 
+def _set_log_file(file_name, worker_pid, old_obj, setter_func):
+    # Line-buffer the output (mode 1).
+    f = open_worker_log(file_name, worker_pid)
+
+    # Before redirecting stdout we need to make sure userspace buffers
+    # are all flushed.
+    # Flush python's userspace buffers
+    old_obj.flush()
```
Comment (ericl) [inline python/ray/worker.py:897]: Have you run into any issues without flushing? Consider removing this.

File: python/ray/worker.py [hunk 25]
```diff
@@ -891,6 +887,71 @@ def custom_excepthook(type, value, tb):
...
+    # and stderr are heavily buffered resulting in seemingly lost logging
+    # statements. We never want to close the stdout file descriptor, dup2 will
+    # close it when necessary and we don't want python's GC to close it.
+    setter_func(open_log(fileno, closefd=False))
+
+    return os.path.abspath(f.name)
+
+
+# lambda cannot contain assignment
+def stdout_setter(x):
```
Comment (ericl) [inline python/ray/worker.py:920]: Can you move this into the function that uses them?

File: python/ray/worker.py [hunk 26]
```diff
@@ -891,6 +887,71 @@ def custom_excepthook(type, value, tb):
...
+        stdout_name (str): The file name that stdout should be written to.
+        stderr_name(str): The file name that stderr should be written to.
+
+    Returns:
+        (tuple) The absolute paths of the files that stdout and stderr will be
+    written to.
+
+    """
+    stdout_path = ""
+    stderr_path = ""
```
Comment (ericl) [inline python/ray/worker.py:942]: Are these used?

File: python/ray/worker.py [hunk 27]
```diff
@@ -1162,37 +1223,17 @@ def connect(node,
...
+                log_stdout_file_name, log_stderr_file_name = \
+                    set_log_file(log_stdout_file, log_stderr_file)
+            except IOError:
+                raise IOError(
+                    "Workers must be able to redirect their output at"
+                    "the file descriptor level.")
     elif not LOCAL_MODE:
         raise ValueError(
             "Invalid worker mode. Expected DRIVER, WORKER or LOCAL.")
+    worker.current_logging_job = None
```
Comment (ericl) [inline python/ray/worker.py:1236]:
```suggestion
    worker.current_job_id = None
```

**Review by ericl — CHANGES_REQUESTED** [review]
Overall looks good

File: python/ray/utils.py (hunk 22 above)
Comment (wuisawesome) [reply python/ray/utils.py:396]:
```
When no buffering argument is given, the default buffering policy works as follows:

Binary files are buffered in fixed-size chunks; the size of the buffer is chosen using a heuristic trying to determine the underlying device’s “block size” and falling back on io.DEFAULT_BUFFER_SIZE. On many systems, the buffer will typically be 4096 or 8192 bytes long.

“Interactive” text files (files for which isatty() returns True) use line buffering. Other text files use the policy described above for binary files.
```
https://docs.python.org/3/library/functions.html#open

File: python/ray/worker.py (hunk 24 above)
Comment (wuisawesome) [reply python/ray/worker.py:897]: I can't find any documentation or guarantee's on python's buffer flushing policy so I'll remove this line but leave a comment. Python does seem to always flush, but this feels like we're relying on very sketchy undefined behavior.

File: python/ray/worker.py (hunk 26 above)
Comment (wuisawesome) [reply python/ray/worker.py:942]: They're passed to `CoreWorker`

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
test_tempfile still seems to be failing, does it pass locally?

**Comment by wuisawesome** [comment]
> test_tempfile still seems to be failing, does it pass locally?

I think you looked at this right before the round of tests finished .
